### PR TITLE
[spirv] Translate for statement

### PR DESCRIPTION
**Pull request by antiagainst**
* Only the basic for statement are tested. More complicated cases
  yet to be implemented and tested.
* Also add support for less than and prefix increment operator.

### CODE REVIEW

**Comment by antiagainst** [comment]
A following up CL will have more complicated test cases for for loops and also fix the basic block ordering for if statement.

File: tools/clang/include/clang/SPIRV/ModuleBuilder.h [hunk 1]
```diff
@@ -97,11 +97,15 @@ class ModuleBuilder {
                           uint32_t rhs);
 
   // \brief Creates an unconditional branch to the given target label.
-  void createBranch(uint32_t targetLabel);
+  // An OpLoopMerge instruction will also be created if both continueLabel
+  // and mergeLabel are not 0.
```
Comment (dneto0) [inline tools/clang/include/clang/SPIRV/ModuleBuilder.h]: What if mergeLabel is nonzero but continueLabel is zero?
If you know that doesn't happen, then say it's a constraint on the caller.

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp [hunk 2]
```diff
@@ -372,12 +374,98 @@ class SPIRVEmitter : public ASTConsumer {
...
+  }
+
   uint32_t doExpr(const Expr *expr) {
     if (auto *delRefExpr = dyn_cast<DeclRefExpr>(expr)) {
       // Returns the <result-id> of the referenced Decl.
       const NamedDecl *referredDecl = delRefExpr->getFoundDecl();
       assert(referredDecl && "found non-NamedDecl referenced");
       return declIdMapper.getDeclResultId(referredDecl);
+    } else if (auto *parenExpr = dyn_cast<ParenExpr>(expr)) {
+      // Juse need to return what's inside the parentheses.
```
Comment (dneto0) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: "Just"

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp [hunk 3]
```diff
@@ -372,12 +374,98 @@ class SPIRVEmitter : public ASTConsumer {
...
+    // The final CFG should normally be like the following. Exceptions will
+    // occur with non-local exits like loop breaks or early returns.
+    //             +--------+
+    //             |  init  |
+    //             +--------+
+    //                 |
+    //                 v
+    //             +--------+
+    //             | header | <----------------+
+    //             +--------+                  |
```
Comment (dneto0) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: Can we always merge header and check blocks?   It seems useless to separate the two.

File: tools/clang/include/clang/SPIRV/ModuleBuilder.h (hunk 1 above)
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/ModuleBuilder.h]: Commented on `createConditionalBranch()`.

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp (hunk 3 above)
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: Good suggestion! That's true. Was literally following the spec. Actually this is a good optimization. Done.

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp (hunk 2 above)
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: Done

**Pull request merged**